### PR TITLE
[flang][debug] Support derived types.

### PR DESCRIPTION
**Pull request by abidh**
This PR adds initial debug support for derived type. It handles `RecordType` and generates appropriate `DICompositeTypeAttr`. The `TypeInfoOp` is used to get information about the parent and location of the derived type.

We use `getTypeSizeAndAlignment` to get the size and alignment of the components of the derived types. This function needed a few changes to be suitable to be used here:

1. The `getTypeSizeAndAlignment` errored out on unsupported type which would not work with incremental way we are building debug support. I have added a new `success` parameter. If it is null, the function behaves same as before and calls `TODO()` for unsupported type. If this is not null, then a true/false status is returned which can be used by clients to see if function succeeded. This is not very clean but gets the job done without duplicating any code.

2. The Character type was returning size of just element and not the whole string which has been fixed.

The testcase checks for offsets of the components which had to be hardcoded in the test. So the testcase is currently enabled on x86_64.

If a derived type has a component whose type is not supported by the `getTypeSizeAndAlignment` then we will generate a place holder type for such derived type.

With this PR in place, this is how the debugging of derived types look like:

```
type :: t_date
    integer :: year, month, day
  end type

  type :: t_address
    integer :: house_number
  end type
  type, extends(t_address) :: t_person
    character(len=20) name
  end type
  type, extends(t_person)  :: t_employee
    type(t_date) :: hired_date
    real :: monthly_salary
  end type
  type(t_employee) :: employee

(gdb) p employee
$1 = ( t_person = ( t_address = ( house_number = 1 ), name = 'John', ' ' <repeats 16 times> ), hired_date = ( year = 2020, month = 1, day = 20 ), monthly_salary = 3.1400001 )
```

### CODE REVIEW

**Review by tblah — COMMENTED** [review]
Style-wise, I would prefer something like
```
std::optional<std::pair<uint64_t, unsigned short>> getTypeSizeAndAlignment(...)
```
So that callers cannot forget to check for the error condition and we do not return anything if nothing is known.

It would be a pain to add the TODO message to every callsite. Maybe putting it in a wrapper called something like `getSizeAndAlignmentOrCrash` which unwraps the optional and prints the TODO message could be one approach.

I think this would be cleaner, but I am open to other ideas.

As for the rest, the code looks good, but please wait for review by somebody more familiar with derived types.

File: flang/test/Transforms/debug-derived-type-1.fir
Comment (abidh) [reply flang/test/Transforms/debug-derived-type-1.fir:3]: The offset of the components of the derived types are hardcoded in the CHECK lines. I was unsure if those will be same for other targets so I restricted the test on x86_64.

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp
Comment (jinisusan) [inline flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:190]: fir::getTypeSizeAndAlignment() does not seem to deal with pointer types (as you have mentioned, we still have unsupported types with which we are dealing in an incremental manner). Hence if the derived type has a pointer member in it (as in the example below), a DICompositeType does not get created for that derived type. Not sure if it would make better sense to create a DICompositeType with whatever members we have discovered so far instead of bailing out. This would provide better debuggability than no DIE getting created eventually for the derived type at all. 
type dtype
    integer :: i
    real :: fi
    character :: ci
    real, pointer :: justaptr
end type dtype

**Review by jinisusan — COMMENTED** [review]
looks good to me other than the comment mentioned.

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp
Comment (abidh) [reply flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:190]: Thanks for the review. The issue I see is that we need the size of the components earlier in the derived type to get the offset of components that come after them. If we don't know the size of any member, it will result in wrong offset of all the components coming after that and debugger will show wrong values for them.

**Comment by abidh** [comment]
> Style-wise, I would prefer something like
> 
> ```
> std::optional<std::pair<uint64_t, unsigned short>> getTypeSizeAndAlignment(...)
> ```
> 
> So that callers cannot forget to check for the error condition and we do not return anything if nothing is known.
> 
> It would be a pain to add the TODO message to every callsite. Maybe putting it in a wrapper called something like `getSizeAndAlignmentOrCrash` which unwraps the optional and prints the TODO message could be one approach.
> 
> I think this would be cleaner, but I am open to other ideas.
> 
> As for the rest, the code looks good, but please wait for review by somebody more familiar with derived types.

I have now changed the function to return an `std::optional` to indicate success/failure. I also added `getTypeSizeAndAlignmentOrCrash` that retains the old behavior to termination compilation if an unsupported type is encountered.

**Review by tblah — APPROVED** [review]
Thanks for the update. Please wait for approval from @jinisusan

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp
Comment (abidh) [reply flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp]: Ah, I understand your point now. I have updated the PR as per your suggestion.

**Review by jinisusan — APPROVED** [review]
LGTM

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp
Comment (jeanPerier) [inline flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:168]: Building a datalayout is not cheap and leverage caching to speed things up (contains many DenseMap, requires doing parsing of module attributes): it should be done only once at the pass level.

https://github.com/llvm/llvm-project/blob/af1d2b9fb1c6065d3a4611f086b8c83a0c05795c/mlir/docs/DataLayout.md?plain=1#L84

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp
Comment (jeanPerier) [inline flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:180]: There will also be an element created for the first component which is the parent component. Does dwarf expect that, or should the parent not be generated as a component?

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp
Comment (jeanPerier) [inline flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:159]: This processing is likely not cheap, do you know how many times this will being called for a given derived type in a program: once per type, or once for every variable with this type, or more?

It may be worth having some caching `fir::RecordType Ty` <-> `mlir::LLVM::DITypeAttr`, or even `mlir::Type` <-> `mlir::LLVM::DITypeAttr`

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp
Comment (jeanPerier) [inline flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:193]: Beware that derived type may be recursive, directly or indirectly with pointer/allocatable components.
You may have to do something special here to prevent infinite loops.

```
module m
 type t1
   type(t2), pointer :: p
 end type
 type t2
   type(t1) :: t1
 end type
 type(t2) :: some_global
end module
```

My guess is that for now you may be OK because getTypeSizeAndAlignment probably returns nullopt for pointer/allocatable component since it does not know how to compute descriptor size, but you may want to add a note and a test to check that this is not causing compiler crashes with -g.

**Review by jeanPerier — COMMENTED** [review]
Thanks!

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp
Comment (abidh) [reply flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:159]: Keeping a cache for `mlir::Type` to `mlir::DITypeAttr` was always the plan. Although I have noticed in the IR dump that there is no type duplication. So may be mlir does caching at some level. I will keep this thing as a TODO to avoid generating redundant types.

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp
Comment (abidh) [reply flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:180]: Yes the base as first argument works. The `DWARF` output is same as generated by classic flang and gfortran. The testcase I added has 3 level of parent-child relationship.

While looking at this, I noticed that I was also setting `base_type` field of the `DICompositeTypeAttr` in this case which is really not needed. That field is for situation where `DICompositeTypeAttr` is representing an array. I will fix it.

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp
Comment (abidh) [reply flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp]: Thanks for the suggestion. I have now updated the code to initialize `DataLayout` once and use it wherever needed.

File: flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp
Comment (abidh) [reply flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:177]: I have added a test for now to make sure we don't crash for such cases. I will handle it when I add support for allocatable components. This is is common issue in C like language so there are ways to handle it.

**Comment by abidh** [comment]
Hi @jeanPerier 
I think I addressed your earlier comments. Please let me know if you have any further comments on this PR.

**Comment by abidh** [comment]
Hi @jeanPerier 
Any further changes required in this PR?

**Comment by abidh** [comment]
Hi @jeanPerier 
I was wondering if you have any further comments on this PR or is it ok to merge? Thanks.

**Review by jeanPerier — APPROVED** [review]
LGTM, thanks

**Pull request merged**
